app/services/provider_interface/sort_application_choices.rb

```diff
@@ -1,24 +1,101 @@
     end
 
-    def self.sort_order(sort_by)
-      return { updated_at: :desc } if sort_by != 'days_left_to_respond'
+    def self.for_task_view(application_choices)
+      application_choices.from <<~WITH_TASK_VIEW_GROUP.squish
+        (
+          SELECT a.*, c.recruitment_cycle_year,
+            CASE
+              WHEN #{deferred_offers_pending_reconfirmation} THEN 1
+              WHEN #{previous_cycle_pending_conditions} THEN 2
+              WHEN #{about_to_be_rejected_automatically} THEN 3
+              WHEN #{awaiting_provider_decision} THEN 4
+              WHEN #{waiting_on_candidate} THEN 5
+              ELSE 999
+            END AS task_view_group,
+            #{pg_days_left_to_respond} AS pg_days_left_to_respond
+
+            FROM application_choices a
+            LEFT JOIN courses c ON c.id = a.course_option_id
+        ) AS application_choices
+      WITH_TASK_VIEW_GROUP
+    end
+
+    def self.deferred_offers_pending_reconfirmation
+      # TODO: this query depends on deferrals implementation
+      <<~DEFERRED_OFFERS_PENDING_RECONFIRMATION.squish
+        (
+          status='deferred'
+            AND c.recruitment_cycle_year = #{RecruitmentCycle.current_year}
+        )
+      DEFERRED_OFFERS_PENDING_RECONFIRMATION
+    end
 
-      Arel.sql(
-        <<-ORDER_BY.strip_heredoc,
+    def self.previous_cycle_pending_conditions
+      <<~PREVIOUS_CYCLE_PENDING_CONDITIONS.squish
         (
-          CASE
-            WHEN (status='awaiting_provider_decision' AND (DATE(reject_by_default_at) > '#{Time.zone.now.iso8601}')) THEN 1
-            ELSE 0
-          END
-        ) DESC,
-        reject_by_default_at ASC,
-        application_choices.updated_at DESC
-        ORDER_BY
-      )
+          status='pending_conditions'
+            AND c.recruitment_cycle_year = #{RecruitmentCycle.previous_year}
+        )
+      PREVIOUS_CYCLE_PENDING_CONDITIONS
+    end
+
+    def self.about_to_be_rejected_automatically
+      <<~AWAITING_PROVIDER_DECISION.squish
+        (
+          status='awaiting_provider_decision'
+            AND c.recruitment_cycle_year = #{RecruitmentCycle.current_year}
+            AND (
+              DATE(reject_by_default_at)
+              BETWEEN
+                DATE('#{Time.zone.now.iso8601}')
+              AND
+                DATE('#{3.business_days.after(Time.zone.now).iso8601}')
+            )
+        )
+      AWAITING_PROVIDER_DECISION
+    end
+
+    def self.awaiting_provider_decision
+      <<~AWAITING_PROVIDER_DECISION.squish
+        (
+          status='awaiting_provider_decision'
+            AND (
+              DATE(reject_by_default_at) >
+                DATE('#{3.business_days.after(Time.zone.now).iso8601}')
+            )
+        )
+      AWAITING_PROVIDER_DECISION
+    end
+
+    def self.waiting_on_candidate
+      <<~WAITING_ON_CANDIDATE.squish
+        (
+          status='offer'
+            AND c.recruitment_cycle_year = #{RecruitmentCycle.current_year}
```

Comment:
I _think_ the offer could have been made in the previous cycle and still require action here?

Reply:
I think you are right :) Some rigorous testing will be needed when we proceed to the implementation, especially around EoC edge cases.